### PR TITLE
[WIP] (#1736) ability to force re-render a story

### PR DESCRIPTION
**Pull request by dangreenisrael**
Issue: #1736

## What I did

added `forceReRender` to the exports in `'@storybook/react'`

## How to test

Storyshots

Does this need a new example in the kitchen sink apps?

Yes

Does this need an update to the documentation?

Yes

### Platforms
- [x] React
- [ ] React Native
- [x] Angular
- [x] Vue

### Functionality
- [x] Ability to re-render from in stories
- [ ] Ability to re-render from addons via channel

### Misc
- [x] Kitchen sink example (cra)
- [x] Screenshots

## Sceenshot
force rerender

## Future possibility 

Standard event that can be emitted to force a reload

### CODE REVIEW

**Comment by danielduan** [comment]
Think we should port this over for vue and angular as well if this was to be added as a feature.

**Comment by dangreenisrael** [comment]
@danielduan  Will do. I'll port it to Vue and Angular when I get a chance, then I'll update the docs for all of the,

File: app/react/src/client/preview/index.js [hunk 1]
```diff
@@ -53,3 +59,5 @@ const renderUI = () => {
 };
 
 reduxStore.subscribe(renderUI);
+
+export const forceReRender = () => renderUI();
```
Comment (ndelangen) [inline app/react/src/client/preview/index.js]: If the story didn't change, why would this work?

Comment (ndelangen) [reply app/react/src/client/preview/index.js]: Never mind, I mis-interpreted the code.

File: app/react/src/client/preview/index.js [hunk 2]
```diff
@@ -20,7 +20,13 @@ const isBrowser =
...
 
 const storyStore = new StoryStore();
-const reduxStore = createStore(reducer);
+/* eslint-disable no-underscore-dangle */
+const reduxStore = createStore(
+  reducer,
+  window.__REDUX_DEVTOOLS_EXTENSION__ &&
+    window.__REDUX_DEVTOOLS_EXTENSION__({ name: 'Storybook Preview', instanceId: 'sbPreview' })
+);
+/* eslint-enable */
```
Comment (tmeasday) [inline app/react/src/client/preview/index.js]: What's happening with this change? It seems unrelated.

File: app/react/src/client/preview/index.js (hunk 1 above)
Comment (tmeasday) [reply app/react/src/client/preview/index.js]: I was just looking at that code, funnily enough and I think it's a bit broken, I was going to send a PR to fix it. 

The current (psuedo) logic is:

```js
render() {
  // line linked above
  if (storyDifferent) {
      ReactDOM.unmountComponentAtNode(rootEl);
  }

  // later
  ReactDOM.render(element, rootEl);
}
```

So it *always* calls `render` and *sometimes* calls `unmount`. This actually breaks the React lifecycle in weird ways.

Comment (tmeasday) [reply app/react/src/client/preview/index.js]: So, if I was to fix the above (probably changing it to: `if (!storyDifferent) return;`, this PR would no longer work.

We should consider if we want to support this `forceRerender` functionality? I suppose we do? If so we will need a more declarative way to make it work, I don't think we should retain the above bug just to make it work.

**Review by tmeasday — CHANGES_REQUESTED** [review]

File: app/react/src/client/preview/index.js (hunk 1 above)
Comment (tmeasday) [reply app/react/src/client/preview/index.js]: Probably the easiest way would be to add a "key" to the redux store that is changed when you call `forceRerender` and to also check if the key has changed.

**Comment by tmeasday** [comment]
@dangreenisrael I am making some of these changes on another branch (because I need the lifecycle to work properly) so LMK if it makes sense to port them across here.

**Comment by tmeasday** [comment]
PS as an aside -- is it expected that all new functionality needs to be implemented in all 3 (should it be 4, what about RN?) view layers to be shipped? For instance, I personally wouldn't have much idea how to (at least idiomatically) implement this in Vue and Angular. It seems like quite a barrier (maybe that's a good thing).

File: app/react/src/client/preview/index.js (hunk 2 above)
Comment (dangreenisrael) [reply app/react/src/client/preview/index.js]: Correct, it is unrelated. It's enabling dev tools. Do you think it should be in a separate branch/pr.

**Comment by dangreenisrael** [comment]
@tmeasday I'm planning to do a deep dive into force re-render over the winter holidays (probably December 27,28,29. Do you think your PR will be merged by then.

**Comment by dangreenisrael** [comment]
@tmeasday @ndelangen Do you guys want to set up a time to chat about force re-render so we can all get on the same page

**Comment by tmeasday** [comment]
@dangreenisrael - https://github.com/storybooks/storybook/pull/2503

I'm pretty convinced we should merge the above. However, I think it shouldn't be too hard to amend the above to consider a `renderKey` or something from the redux store.

**Comment by dangreenisrael** [comment]
@tmeasday, I'll rebase off 3.3/your branch before getting back to work on this

**Comment by ndelangen** [comment]
Please check if this still works after @tmeasday's change..

**Comment by tmeasday** [comment]
Changes look good to me.

File: app/react/src/client/preview/index.js (hunk 2 above)
Comment (tmeasday) [reply app/react/src/client/preview/index.js]: I think yeah, let's move to a separate PR

**Review by tmeasday — DISMISSED** [review]
LGTM, let's get rid of the devtools

**Comment by ndelangen** [comment]
merge conflicts

**Comment by dangreenisrael** [comment]
Merging as @tmeasday's requested changes were made

**Comment by stefanKuijers** [comment]
Thank you guys for the great work. As far as I can see this has been merged in and master.  I'm using version 3.3.12 but cannot run the [example code](https://github.com/storybooks/storybook/blob/82699d3f0d98d897e2cd8316135aeb45fb4d85c5/examples/cra-kitchen-sink/src/stories/force-rerender.stories.js).

@dangreenisrael: Sadly when using the example code I run into the following error
When I call ```forceReRender``` I get the following error:
```Uncaught TypeError: Object(...) is not a function
    at increment (0.345baf011dfa0388ca2e.hot-update.js:70)
    at HTMLUnknownElement.callCallback (preview.bundle.js:30287)
    at Object.invokeGuardedCallbackDev (preview.bundle.js:30326)
    at Object.invokeGuardedCallback (preview.bundle.js:30183)
    at Object.invokeGuardedCallbackAndCatchFirstError (preview.bundle.js:30197)
    at executeDispatch (preview.bundle.js:30581)
    at executeDispatchesInOrder (preview.bundle.js:30603)
    at executeDispatchesAndRelease (preview.bundle.js:30701)
    at executeDispatchesAndReleaseTopLevel (preview.bundle.js:30712)
    at Array.forEach (<anonymous>)
```
It is thrown two times: ```0.345baf011dfa0388ca2e.hot-update.js:70``` & @ ```preview.bundle.js:30355```. 

When I look into the [cra-kitchensink folder](https://github.com/storybooks/storybook/tree/82699d3f0d98d897e2cd8316135aeb45fb4d85c5/examples/cra-kitchen-sink) I don't see anything special in the setup which differs from mine. Is there still a change this is related to setup? For example that I'm using different loaders? (TS, SCSS)

Thanks in advance!

**Comment by dangreenisrael** [comment]
@stefanKuijers could you open up an issue for this? Also, can you reproduce this issue in a public repo?

**Comment by stefanKuijers** [comment]
Thanks for the reply guys. I saw a few times in the branch names a 3.3 which is why I thought it was still part of 3.3. 

Thanks for the great work! I will wait until 3.4.0. I see on the side now that this is assigned to the 3.4 milestone.
